types/progress-stream/progress-stream-tests.ts

```diff
@@ -0,0 +1,71 @@
+import progress = require("progress-stream");
+import stream = require("stream");
+
+const options: progress.Options = {
+    time: 100,
+    speed: 100,
+    length: 100,
+    drain: true,
+    transferred: 0,
+};
+
+const progressListener = (progress: progress.Progress) => {
+    // $ExpectType number
+    progress.percentage;
+    // $ExpectType number
+    progress.transferred;
+    // $ExpectType number
+    progress.length;
+    // $ExpectType number
+    progress.remaining;
+    // $ExpectType number
+    progress.eta;
+    // $ExpectType number
+    progress.runtime;
+    // $ExpectType number
+    progress.delta;
+    // $ExpectType number
+    progress.speed;
+};
+
+// $ExpectType ProgressStream
+const p = progress();
+
+// $ExpectType ProgressStream
+progress(options);
+
+// $ExpectType ProgressStream
+progress(options, progressListener);
+
+// $ExpectType ProgressStream
+progress(progressListener);
+
+// $ExpectType ProgressStream
+p.on("progress", progressListener);
+
+// $ExpectType ProgressStream
+p.on("length", (length: number) => {});
+
+p.setLength(200); // $ExpectType void
+
+p.progress(); // $ExpectType Progress
+
+// Check if ProgressStream extends stream.Transform correctly
+
+// $ExpectType ProgressStream
```

Comment:
These tests below are failing when running `npm run test` and when running `dtslint` manually.

It seems to think that these methods return a `stream.Transform` instead of a `ProgressStream`.

![2019-02-16_11-12-09](https://user-images.githubusercontent.com/5085260/52898652-7660c700-31e0-11e9-90a6-2253189226ae.png)

TypeScript@3.3.3 does not report any errors when compiling code that uses these definitions (nor any problems in vscode). TSLint (using vscode-typescript-tslint-plugin@1.0.0) does not report any problems with the definitions or with the tests.

![2019-02-16_11-45-12](https://user-images.githubusercontent.com/5085260/52898636-59c48f00-31e0-11e9-85d1-375e0c2c0cf6.png)


Both of these interpret the return value as `ProgressStream`. If I change the `$ExpectType` assertion to expect a `Transform` instead, TSLint says the assertion failed because it got a `ProgressStream`.

![2019-02-16_11-42-40](https://user-images.githubusercontent.com/5085260/52898622-31d52b80-31e0-11e9-9a46-2c4ed41a2d6a.png)

Does anyone know what's going on here?

Reply:
@mickdekkers, thanks for following up!  Something's definitely wrong, since the language service and linter disagree, but I'm not sure what the correct behavior is.  I've filed https://github.com/Microsoft/TypeScript/issues/30031.

In the meantime, I'd suggest using `extends`, rather than an intersection type, to work around this problem.  Sorry for the trouble.

Reply:
Hi @amcasey, thanks for looking into this!

The reason I used an intersection type is because I could find no way of extending `stream.Transform` with new overloads for `on(event)`.

Specifically, I initially tried to define `ProgressStream` as an interface like this:

```ts
interface ProgressStream extends stream.Transform {
    on(event: "progress", listener: ProgressListener): ProgressStream;
    on(event: "length", listener: (length: number) => void): ProgressStream;
    setLength(length: number): void;
    progress(): Progress;
}
```

But this way TypeScript reports a compile error (using `npm run lint progress-stream`):

```
Error: /DefinitelyTyped/types/progress-stream/index.d.ts:34:15
ERROR: 34:15  expect  TypeScript@next compile error:
Interface 'ProgressStream' incorrectly extends interface 'Transform'.
  Types of property 'on' are incompatible.
    Type '{ (event: "progress", listener: ProgressListener): ProgressStream; (event: "length", listener: (length: number) => void): ProgressStream; }' is not assignable to type '{ (event: "close", listener: () => void): this; (event: "data", listener: (chunk: any) => void): this; (event: "end", listener: () => void): this; (event: "readable", listener: () => void): this; (event: "error", listener: (err: Error) => void): this; (event: string | symbol, listener: (...args: any[]) => void): thi...'.
      Types of parameters 'event' and 'event' are incompatible.
        Type '"close"' is not assignable to type '"progress"'.

/DefinitelyTyped/types/progress-stream/progress-stream-tests.ts:56:6
ERROR: 56:6   expect  TypeScript@next compile error:
Argument of type '"close"' is not assignable to parameter of type '"length"'.
ERROR: 58:6   expect  TypeScript@next compile error:
Argument of type '"data"' is not assignable to parameter of type '"length"'.
ERROR: 60:6   expect  TypeScript@next compile error:
Argument of type '"end"' is not assignable to parameter of type '"length"'.
ERROR: 62:6   expect  TypeScript@next compile error:
Argument of type '"error"' is not assignable to parameter of type '"length"'.
ERROR: 64:6   expect  TypeScript@next compile error:
Argument of type '"readable"' is not assignable to parameter of type '"length"'.
```

Note that `close`, `data`, `end`, `error`, and `readable` are all events on `stream.Transform`. (These event definitions are part of `stream.Readable`, which is extended by `stream.Duplex`, which in turn is extended by `stream.Transform`.)

I [asked for advice](https://gitter.im/Microsoft/TypeScript?at=5c67247b604f233ab617c4e7) on the TypeScript Gitter, which is what led me to the current definitions using an intersection type.

Do you know of any other way to do this?

Reply:
Hi @mickdekkers, to clarify, the way you're doing it now makes perfect sense.  I only suggested changing it as a way of working around a bug.  To get the extends version to compile, you will, unfortunately, have to _redeclare_ all of the overloads from `Transform`.  Your way is better, but this will make it possible to merge before the bug is fixed. 

I will, somewhat reluctantly, point out that you also have the option of adjusting your test.  If you do so, please leave a comment explaining why.

Reply:
Hi @amcasey, thanks for getting back to me! I was hoping to avoid redeclaring all the overloads, but it does look like the `extends` version is the best option for now 🙂

I updated the `ProgressStream` definition and added an explanation in a comment.